### ISSUE TITLE
USB Drivers

### ISSUE THREAD

**Issue by Lesinh257** [WOA-Project/SurfaceDuo-Guides#74]
It's great that you guys are working hard to get windows working on surface duo 1, unfortunately the usb type C port on my surface doesn't support peripherals, I hope you can improve this as soon as possible. Thanks for your efforts for the community

**Comment by etienne51** [comment]
Asking the same thing here. I know there has been a lot of progress being made on many other things, congrats on everything, but to me this, and USB charging has been a complete deal breaker from the start. I'm very interested in this Windows port, but it's not gonna see any use at all for me until USB charging under Windows and USB peripherals are supported. Charging itself seems like the absolute most important thing of all, especially considering how fast the battery is drained under Windows, so I originally thought this would be the top priority, but update after update, it's still in the same non working state. I seem to remember it was barely starting to work like a whole year ago or something, but support has been dropped again after that. Really hoping this is going to be supported soon. In the meantime, my Surface Duo 1 goes back to the drawer for a few more months (not my main phone, I was just curious about that Windows thing, so I just got it to tinker with).

**Comment by Lesinh257** [comment]
It's great now, I just hope the usb type c port gets full functionality and the speakers work then the surface duo 1 is a device that can replace bulky laptops when I leave the house

**Comment by gus33000** [comment]
This message is going to be quite lengthy; because there are multiple things in this thread to address (even if for end users it may not seem like it), I'm sorry in advance for the length.

I do want to preface this with that i appreciate the interest here; I get your complaints and they are valid. And I am thankful you are providing this feedback too (wish more people would in case of issues they come across). And I am not saying these issues dont exist either. I appreciate the feedback a lot but i am also providing a bit more detail because some of these functionalities are meant to be fully working today.

---

Regarding USB Peripherals such as _mice, keyboards_ plugged into the device directly; using a USB-C to USB-A or USB-C Hub (Hub that **does not take external power in from the wall**):

This is a known issue regarding the Power Delivery State Machine getting stuck between alert messages and there is no fix being worked on as we speak at the moment.

---

Regarding USB Peripherals such as _mice, keyboards_ plugged into the device directly; using a USB-C to USB-A or USB-C Hub (Hub that **does take external power in from the wall**):

This is meant to be fully working **today**. The only caveat is that it will break the very moment you swap the USB cable with another one and try again. It will only work for the very first insertion since OS boot; the reason this is the case is similar to the dongle issue from above.

---

Regarding DP Alt Mode over USB-C:

State machine issue similar to above for why

---

Regarding charging:

Much like every issue mentioned above; it only works once; otherwise the state machine is stuck, in a very similar fashion.

![PXL_20241215_193913285](https://github.com/user-attachments/assets/d15555dc-1424-496c-ac3f-e5524188ddea)

Here the device charges fine and increases fine as well, alebit slowly (because I am plugged in into a standard PC USB port).

I booted the OS; with no cable plugged in, and inserted the cable, then it started working (will be the same with any PD charger). The moment I however disconnect said cable; do something else; and replug it; its not gonna work; and will decrease like you are all mentioning.

I do confess there is a problem with the speed yes, but it is meant to be working fine.

The absolute priority when working on this was **safety** and not making charging work. This device uses a dual battery pack system with no bms; all in software; and with components originally designed for fast battery pack charging in parallel; hacked together to create an asymetric charging system that must be kept track of every 30 seconds to ensure a proper balance between both packs SOCs as well as no more than 1A current going through the hinge wires (or they would burn). These safety measures are also the reason it is still slow today as I lack the proper equipment to make sure this is fully working and i can change the settings so it works fine. But it charges fine still; even if slower.

The support and status page also says the same thing; that charging is working. But i believe the confusion comes from the bug that does not allow anything to work once you do a single cable swap.

---

As for speakers out; no ETA nor is this functional yet. I am sorry but you may have to wait more.

---

In general I do believe this is one and the same issue being described here, because most of this aside from DP over USB-C and Dongles is meant to be working fine.

I get the furstration though but given I am working alone on this project and things have been getting more, filled this year for me with other IRL things; there have been higher priorities so far; especially when I dont consider these issues to be deal breakers either.

I dont want to commit to any ETA because none of this is simple nor fully known or worked out for how to make it work properly; and while this may sound painful to hear especially when some of you have been waiting for over a year for these problems to be ironed out; i did make the choice to focus on other things this entire year to fix; address and enable.

These things may not matter to you but we have been doing lots either way. And I dont see much that cant be worked around.

I appreciate the enthusiam so far here but I cannot also say that yeah these problems have been fixed and will go out soon; because this isnt the case and it really isnt as simple as you may think these are to address.

Safety and stability will always come above functionality and new features, for many reasons (especially when a lot of hardware components are interconnected in such a way that a problem in a specific area may impact lots (the state machine issue so far is impacting 6 different features that end users would consider separate; as you can see in the short list of 4 things right above). I also dont want to introduce new functionality in bulk to then get a very very buggy and broken experience; with no protection; that keeps crashing; and damages end user devices (which also cost their own sum of money on their own). Even if i didnt mind about breaking end user devices, fixing such issues would become exponentially more complicated as more issues arise.

At least for now you should have workarounds, but these core hardware features are meant to be working more or less. Non functional is def odd. I would be curious either way to know if its really straight up non functional for all of you, and if it is; in such case; would appreciate more detail on what charger (amazon link for example or model?) or peripheral is causing problems; so we can debug this further, as the USB spectrum is huge; and the USB specification is equally insane.

**Comment by gus33000** [comment]
USB problems in particular are also annoying to fix and debug; partly due to the only way to reliably debug the operating system is via the USB port and when doing so; the USB port is reserved for use by the kernel debugger and we of course cannot debug issues in this case.

**Comment by gus33000** [comment]
![PXL_20241215_200108761](https://github.com/user-attachments/assets/df84ff41-67a7-400e-bace-1088e49ffa27)

just for reference; the ETA for full charge might be a little skewed as well; because the device is now at 97% ever since my original message too. So maybe I also need to look into this. It def wasnt half an hour

**Comment by etienne51** [comment]
Hi Gus, thank you for your very thorough answer!

I tested things a bit, and indeed I have USB peripherals working perfectly fine using an Icy Box IB-DK4050CPD hub, with PD power plugged in to it externally. Mouse and keyboard attached, both working, and later I plugged a USB drive and it also shows up fine as long as the hub remains plugged in. Great!

Now regarding charging. This is another issue. I'm plugged in to a proper 140W PD3.1 wall adapter, via the hub. I have a power analyzer in line, which shows 12V has been selected, but I only get 2.77W which is not enough for my Surface Duo to even maintain it's charge in Windows. I don't have the small lightning logo on the battery in the taskbar either and the popup says "Battery status: 83% remaining" as I'm writing here, no mention of charging sadly.

I find this a bit weird, because when using the Surface Duo in bootloader mode, or in TWRP, it always requests 5V (if I remember right), and then charges at around 10W, maybe a bit more, sometimes I've seen 13W I think. I've only seen that 12V input here in Windows, but though only with 2.77W of power, which is definitely insufficient.

Regarding the version of Windows and drivers I have, I literally just got back to tinkering with my Duo a couple days ago, and fully reinstalled from scratch, removing the old Windows, upgrading Android to the latest version, and then using WOA Device Manager from my desktop PC to flash the latest FFU of v2411.80 (Nov. 24, 2024). That's it.

I have a Clockwork uConsole around, I'm thinking of pulling the keyboard/trackball and 3D printing something to use it with the Surface Duo maybe, just an idea!

**Comment by etienne51** [comment]
![DSC01491_edited_resized](https://github.com/user-attachments/assets/3b97b88a-6e95-452f-ab36-67640df46a02)

Well, I had the power meter attached before the hub instead of after. What I was seeing was just the power draw from the peripherals plugged in. Unfortunately, the Surface Duo itself, while it requests 12V itself, does not draw any current at all on Windows.

My Android does not seem to want to boot anymore, I'll see if I can figure out why. I don't know what voltage it requests under Android, but I know it's 5V in the bootloader, and it pulls around 8-13W. Maybe I will test on the USB port of my PC, which is also PD but only 5V at 3A (15W) max. So that will lock to 5V, it won't be able to request 12V, and maybe it will charge that way? I'll comment back after testing that.

**Comment by gus33000** [comment]
No lightning symbol == we didn't detect you inserted a cable, it's not gonna charge as a result there's no way even if you force power into it. No point in analysing the draw as a result, as it's unaware of a cable being inserted. This sounds like either a bad pd implementation that got rejected or the stuck state machine. Try another input. I will also mention this type of current analyser can mislead the state machine

**Comment by etienne51** [comment]
So I did some more testing. I have to say for me charging seems completely random. You're right, it looks like with the power meter inline it can cause issues, I don't seem to remember being able to charge with it connected at all. But it's quite random so I'm not sure.

Without the power meter, meaning a simple USB C cable from the charger to the hub, and then the hub directly connected to the phone, I had a really hard time getting the phone to charge!

After booting into Windows and reaching the desktop, I plugged in the hub (it was already plugged in to the charger), and sometimes I had the mouse and keyboard working, sometimes not. When the mouse and keyboard did not work, generally unplugging and replugging the hub once solved the issue.

For charging though, only once I had charging working right away. Another time I had to unplug and replug 3 times, and on the third time only the lightning icon stuck and charging was working, with the battery percentage slowly going up. Some other times, despite the icon showing up, not even using the phone at all (on the desktop, no apps running), the battery percentage kept going down slowly. Very often, I have to fight, unplug/replug many many times to get the lightning icon to stick, once it was probably like 15 times, I lost count.

The thing I noticed is, when plugging in the hub, most times the lightning icon will blip for a second and go away afte that. When that happens, it means I can unplug and replug the cable without loosing hub functionnality. So I try again... same thing, etc... until either charging works, or the lightning icon stops showing up alltogether. If it stops showing completely, I also lost the USB peripherals, and I have to reboot and retry. Sometimes I had to reboot multiple times until I got charging to finally work.

I tested with both my PC (outputing 5V) and my wall charger (cappable of 12V) I managed to get charging to work on both at least once if I'm not mistaken.

**Comment by gus33000** [comment]
What you describe sounds like the state machine issue but you're also not trying to boot with what you want to try already plugged into the port. That's the sole scenario that at least for me is flawless.

**Comment by etienne51** [comment]
Well, I should have tested that too, that was dumb.

So when booting with the hub already plugged in, it reliably shows the charging indicator pretty much every single time. That said, it remains quite finicky for me.

It looks like charging from my computer USB C with it's 5V only output "works", though it seems to have barely enough power to maintain the battery percentage (I didn't test for long though).

I tried wall chargers, and well... I first used a power meter to see what was going on, it does switch to fast charging at 12V, and pulls around 18W or so. Big issue though, with a charger cappable of 12V output, after maybe 30s or maybe more at times, I get a BSOD! I tried without the power meter, but it's the exact same thing. Once, on my power bank it did not throw a BSOD, but the power bank was showing 3W, and the Surface Duo battery percentage was decreasing. After a couple minutes, all USB peripherals just randomly disconnected, and I had to manually reboot.

So I tried another charger that only outputs 9V max (I think that was my Google Pixel charger or something). Well... this time it does fast charge, and no BSOD at 9V... but every few seconds all peripherals just disconnects and reconnects, and at some point it just stays disconnected and won't charge anymore. So again, manual reboot needed when that happens.

I remember there was a combinaison that worked at some point, and I was seeing the battery percentage go up, but after going away for a few minutes, when I got back it had rebooted (probably a BSOD I guess). I forgot what charger I used for that test.

Anways, after all that testing, the battery is low, so I'm charging from TWRP right now (charges at around 7W), and I will do some more testing later.

**Comment by gus33000** [comment]
Thanks a lot for all of these details as theyre very useful too for when Ill have more time to focus more on the state machine code here!

---

The bugcheck might be from the battery protection code itself (we bugcheck the system in the case of a non safe value condition being detected that could harm the hinge wires or one of the battery packs) but without the minidump i can only guess that theory (they should be under c:\windows\minidumps).

**Comment by etienne51** [comment]
Alright, somehow after fully charging outside of Windows, now my battery is around 89% and it managed to not BSOD anymore running off my powerbank outputing 12V, around 10W (probably 3W to the peripherals, so around 7W to the Duo itself). At bootup it was 22W from the powerbank, but quickly dropped around 14W and then 10W sustaining now.

![DSC01508_edited](https://github.com/user-attachments/assets/f2a19fc7-c1fa-4ee3-a97a-d21e0319b227)

I launched a game (WoW Classic runs at a very respectable 40-50fps on that thing, since it's native ARM!), and it seems to very very slowly drain the battery, despite it being plugged in. I even saw the battery rise to 90% after maybe 5min in the game, but now after maybe 15min it's down to 88%. Quite reasonnable, though there is a lot of margin for the power input, with only around 10W at the moment. With proper power management, the battery should charge itself even with demanding software running at the same time.

![DSC01500](https://github.com/user-attachments/assets/874c41a2-98ed-4016-9ced-8bed54d5a240)

Here is the last dump file I have, if you need previous ones as well let me know. Also, if you need me to do some testing later on when you work on the power management, I might be able to help.

[121724-22453-01.dmp](https://github.com/user-attachments/files/18172390/121724-22453-01.dmp)
